x-pack/filebeat/input/internal/httpmon/roundtripper.go

```diff
@@ -0,0 +1,194 @@
+		reqGet:        monitoring.NewUint(reg, "http_request_get_total"),
+		reqHead:       monitoring.NewUint(reg, "http_request_head_total"),
+		reqOptions:    monitoring.NewUint(reg, "http_request_options_total"),
+		reqPatch:      monitoring.NewUint(reg, "http_request_patch_total"),
+		reqPost:       monitoring.NewUint(reg, "http_request_post_total"),
+		reqPut:        monitoring.NewUint(reg, "http_request_put_total"),
+		reqsAccSize:   monitoring.NewUint(reg, "http_request_body_bytes_total"),
+		reqsSize:      metrics.NewUniformSample(1024),
+		resps:         monitoring.NewUint(reg, "http_response_total"),
+		respErrs:      monitoring.NewUint(reg, "http_response_errors_total"),
+		resp1xx:       monitoring.NewUint(reg, "http_response_1xx_total"),
+		resp2xx:       monitoring.NewUint(reg, "http_response_2xx_total"),
+		resp3xx:       monitoring.NewUint(reg, "http_response_3xx_total"),
+		resp4xx:       monitoring.NewUint(reg, "http_response_4xx_total"),
+		resp5xx:       monitoring.NewUint(reg, "http_response_5xx_total"),
+		respsAccSize:  monitoring.NewUint(reg, "http_response_body_bytes_total"),
+		respsSize:     metrics.NewUniformSample(1024),
+		roundTripTime: metrics.NewUniformSample(1024),
+	}
+
+	_ = adapter.GetGoMetrics(reg, "http_request_body_bytes", adapter.Accept).
+		GetOrRegister("histogram", metrics.NewHistogram(out.reqsSize))
+	_ = adapter.GetGoMetrics(reg, "http_response_body_bytes", adapter.Accept).
+		GetOrRegister("histogram", metrics.NewHistogram(out.respsSize))
+	_ = adapter.GetGoMetrics(reg, "http_round_trip_time", adapter.Accept).
+		GetOrRegister("histogram", metrics.NewHistogram(out.roundTripTime))
+
+	return out
+}
+
+// RoundTrip implements the http.RoundTripper interface, sending
+// request and response metrics to the underlying registry.
+//
+//	http_request_total
+//	http_request_errors_total
+//	http_request_delete_total
+//	http_request_get_total
+//	http_request_head_total
+//	http_request_options_total
+//	http_request_patch_total
+//	http_request_post_total
+//	http_request_put_total
+//	http_request_body_bytes_total
+//	http_request_body_bytes
+//	http_response_total
+//	http_response_errors_total
+//	http_response_1xx_total
+//	http_response_2xx_total
+//	http_response_3xx_total
+//	http_response_4xx_total
+//	http_response_5xx_total
+//	http_response_body_bytes_total
+//	http_response_body_bytes
+//	http_round_trip_time
+func (rt *MetricsRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if rt.metrics == nil {
+		return rt.transport.RoundTrip(req)
+	}
+
+	rt.metrics.reqs.Add(1)
+
+	rt.monitorByMethod(req.Method)
+
+	if req.ContentLength >= 0 {
+		rt.metrics.reqsAccSize.Add(uint64(req.ContentLength))
+		rt.metrics.reqsSize.Update(req.ContentLength)
+	}
+
+	reqStart := time.Now()
+	resp, err := rt.transport.RoundTrip(req)
+	rt.metrics.roundTripTime.Update(time.Since(reqStart).Nanoseconds())
+
+	if resp != nil {
+		rt.metrics.resps.Add(1)
+	}
+
+	if resp == nil || err != nil {
```

Comment:
any specific reason why we are not checking this immediately after the error is returned ?

Reply:
This is all existing code with no semantic change. It was added in #35392.